ModLoader/ModFilesystem.cpp

```diff
@@ -0,0 +1,194 @@
+#include "ModFilesystem.h"
+
+#include <algorithm>
+#include <ranges>
+
+
+#include "../OSCompatibilityLayer.h"
+
+
+
+namespace
+{
+
+bool PathIsReplaced(std::string_view path, const std::set<std::string>& replaced_paths)
+{
+	return std::ranges::any_of(replaced_paths, [&path](const std::string& replaced_path) {
+		if (path.starts_with(replaced_path))
+		{
+			return true;
+		}
+		if (!replaced_path.starts_with('/') && path.starts_with('/' + replaced_path))
+		{
+			return true;
+		}
+		return false;
+	});
+}
+
+} // namespace
+
+
+std::optional<std::string> commonItems::ModFilesystem::GetActualFileLocation(const std::string& path) const
+{
+	for (const auto& mod: mods_ | std::views::reverse)
+	{
+		if (DoesFileExist(mod.path + '/' + path))
+		{
+			return mod.path + '/' + path;
+		}
+		if (PathIsReplaced(path, mod.replacedFolders))
+		{
+			return std::nullopt;
+		}
+	}
+
+	// check game root
+	if (DoesFileExist(game_root_ + '/' + path))
+	{
+		return game_root_ + '/' + path;
+	}
+
+	return std::nullopt;
+}
+
+
+std::optional<std::string> commonItems::ModFilesystem::GetActualFolderLocation(const std::string& path) const
+{
+	for (const auto& mod: mods_ | std::views::reverse)
+	{
+		if (DoesFolderExist(mod.path + '/' + path))
+		{
+			return mod.path + '/' + path;
+		}
+		if (PathIsReplaced(path, mod.replacedFolders))
+		{
+			return std::nullopt;
+		}
+	}
+
+	// check game root
+	if (DoesFolderExist(game_root_ + '/' + path))
+	{
+		return game_root_ + '/' + path;
+	}
+
+	return std::nullopt;
+}
+
+
+std::set<std::string> commonItems::ModFilesystem::GetAllFilesInFolder(const std::string& path) const
+{
+	std::set<std::string> full_files;
+	std::set<std::string> found_files;
+
+	for (const auto& mod: mods_ | std::views::reverse)
+	{
+		for (const auto& new_file: commonItems::GetAllFilesInFolder(mod.path + '/' + std::string(path)))
+		{
+			if (found_files.contains(new_file))
+			{
+				continue;
+			}
+
+			found_files.insert(new_file);
+			full_files.insert(mod.path + '/' + path + '/' + std::string(new_file));
```

Comment:
Hmm. I believe you're inserting the first match in a mod that contains this file and ignoring further instances. However if a second mod is a compatch that has current mod as a dependency then it should override this file with corrected one.

Reply:
Do we really need to handle compatches in a special way? A compatch should be last in the load order, so if we load the mods in load order and iterate over that but reversed, we will get the overridden file from compatch mod, no?

Reply:
No, that't not the case, unfortunately. Apparently, at least for older games like EU4 and HoI4, the load order provided by the launcher does absolutely nothing, the game sorts the load order internally according to dependencies listed in the .mod file.

But we don't even see the mod load order, we grab all the files from disk alphabetically, and with ugc_<crap>.mod filenames mod order is in fact completely random.

Reply:
WTF why would they do that. I don't want to touch a non-Jomini game ever again.

Reply:
I guess we could add an `is_jomini` parameter to `ModFilesystem` constructor, and depending on that use the load order logic or the older game dependencies "logic".

Reply:
Or just drop older games, EU5 is probably coming relatively soon 😈

Reply:
The intent was that the caller passes in the mods in the intended order. So it's up to the caller to sort based on dependencies. I'll add a function to the mod loader in another PR after merging this.